kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java

```diff
@@ -0,0 +1,166 @@
+package com.slack.kaldb.chunk.manager;
+
+import brave.ScopedSpan;
+import brave.Tracing;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.linecorp.armeria.common.RequestContext;
+import com.slack.kaldb.blobfs.s3.S3BlobFs;
+import com.slack.kaldb.blobfs.s3.S3BlobFsConfig;
+import com.slack.kaldb.chunk.Chunk;
+import com.slack.kaldb.logstore.search.SearchQuery;
+import com.slack.kaldb.logstore.search.SearchResult;
+import com.slack.kaldb.logstore.search.SearchResultAggregator;
+import com.slack.kaldb.logstore.search.SearchResultAggregatorImpl;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.spotify.futures.CompletableFutures;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ChunkManager<T> extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(ChunkManager.class);
+
+  protected final Map<String, Chunk<T>> chunkMap = new ConcurrentHashMap<>();
+
+  // TODO: We want to move this to the config eventually
+  // Less than KaldbDistributedQueryService#READ_TIMEOUT_MS
+  public static final int QUERY_TIMEOUT_SECONDS = 10;
+  public static final int LOCAL_QUERY_THREAD_POOL_SIZE = 4;
+
+  private static final ExecutorService queryExecutorService = queryThreadPool();
+
+  /*
+     One day we will have to think about rate limiting/backpressure and we will revisit this so it could potentially reject threads if the pool is full
+  */
+  private static ExecutorService queryThreadPool() {
+    return Executors.newFixedThreadPool(
+        LOCAL_QUERY_THREAD_POOL_SIZE,
+        new ThreadFactoryBuilder().setNameFormat("chunk-manager-query-%d").build());
+  }
+
+  /*
+   * Query the chunks in the time range, aggregate the results per aggregation policy and return the results.
+   * We aggregate locally and and then the query aggregator will aggregate again. This is OKAY for the current use-case we support
+   * 1. topK results sorted by timestamp
+   * 2. histogram over a fixed time range
+   * We will not aggregate locally for future use-cases that have complex group by etc
+   */
+  public CompletableFuture<SearchResult<T>> query(SearchQuery query) {
+
+    SearchResult<T> errorResult =
+        new SearchResult<>(new ArrayList<>(), 0, 0, new ArrayList<>(), 0, 0, 1, 0);
+
+    List<CompletableFuture<SearchResult<T>>> queries =
```

Comment:
this might qualify for a long chain award in terms of LOC :) Anyways this was always there

Reply:
https://github.com/slackhq/kaldb/pull/56/files#diff-cf8f5116a98917009ffaee10b4534d5695a235497100e64eab54d90937543d40R283 😄